mmv1/third_party/terraform/fwvalidators/framework_validators.go

```diff
@@ -0,0 +1,288 @@
+	return "value must be a valid service account email address"
+}
+
+func (v ServiceAccountNameValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+func (v ServiceAccountNameValidator) ValidateString(ctx context.Context, req validator.StringRequest, resp *validator.StringResponse) {
+	if req.ConfigValue.IsNull() || req.ConfigValue.IsUnknown() {
+		return
+	}
+
+	value := req.ConfigValue.ValueString()
+	valid := false
+	for _, pattern := range serviceAccountNamePatterns {
+		if matched, _ := regexp.MatchString(pattern, value); matched {
+			valid = true
+			break
+		}
+	}
+
+	// Check for empty string
+	if value == "" {
+		resp.Diagnostics.AddError("Invalid Service Account Name", "Service account name must not be empty")
+		return
+	}
+
+	if !valid {
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			"Invalid Service Account Name",
+			"Service account name must match one of the expected patterns for Google service accounts",
+		)
+	}
+}
+
+// Create a custom validator for duration
+type DurationValidator struct {
+}
+
+func (v DurationValidator) Description(ctx context.Context) string {
+	return "value must be a valid duration string less than or equal to 1 hour"
+}
+
+func (v DurationValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+func (v DurationValidator) ValidateString(ctx context.Context, req validator.StringRequest, resp *validator.StringResponse) {
+	if req.ConfigValue.IsNull() || req.ConfigValue.IsUnknown() {
+		return
+	}
+
+	value := req.ConfigValue.ValueString()
+	duration, err := time.ParseDuration(value)
+	if err != nil {
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			"Invalid Duration Format",
+			"Duration must be a valid duration string (e.g., '3600s', '1h')",
+		)
+		return
+	}
+
+	if duration > 3600*time.Second {
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			"Duration Too Long",
+			"Duration must be less than or equal to 1 hour",
+		)
+	}
+}
+
+// ServiceScopeValidator validates that a service scope is in canonical form
+var _ validator.String = &ServiceScopeValidator{}
+
+// ServiceScopeValidator validates service scope strings
+type ServiceScopeValidator struct {
+}
+
+// Description returns a plain text description of the validator's behavior
+func (v ServiceScopeValidator) Description(ctx context.Context) string {
+	return "service scope must be in canonical form"
+}
+
+// MarkdownDescription returns a markdown formatted description of the validator's behavior
+func (v ServiceScopeValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+// ValidateString performs the validation
+func (v ServiceScopeValidator) ValidateString(ctx context.Context, req validator.StringRequest, resp *validator.StringResponse) {
+	if req.ConfigValue.IsNull() || req.ConfigValue.IsUnknown() {
+		return
+	}
+
+	canonicalized := CanonicalizeServiceScope(req.ConfigValue.ValueString())
```

Comment:
[CanonicalizeServiceScope from the tpgresource package](https://github.com/GoogleCloudPlatform/magic-modules/blob/3db026ca076c34e112178ab9c5e5a6a7fb251e8b/mmv1/third_party/terraform/tpgresource/service_scope.go#L5) is a helper to convert shorthand strings into the long version of the scope's name that is API-compatible. In the google_service_account_access_token data source it is [used to change the value stored in state](https://github.com/hashicorp/terraform-provider-google/blob/45df35f12441f4d6749a76f35d26fd89647aa941/google/services/resourcemanager/data_source_google_service_account_access_token.go#L39-L41), not validate the input from users.

`ServiceScopeValidator` currently only flags strings as invalid if they are represented in CanonicalizeServiceScope's map, so I think the validator creates a false sense of security. E.g. "foobar" is found to be valid but "bigquery" isn't.

If we want to enforce that all the inputs are a canonical/long-form scope name then I think we should use something like a regex to check the format of the input strings. We could include the use of CanonicalizeServiceScope to allow the convenience of users supplying short names for some scopes, but I think the output of CanonicalizeServiceScope should be checked using the regex after. [I've asked in the shared chat if there are any gotchas using regex on scopes.
](https://chat.google.com/room/AAAA49vo02s/Cvatj3mIFfM/Cvatj3mIFfM?cls=10)

An alternative, instead of updating the validator, is that we could skip implementing a validator for scopes because scopes have not been validated in the provider previously. What do you think?

Reply:
From the discussion I linked it looks like [validating scopes isn't possible](https://chat.google.com/room/AAAA49vo02s/Cvatj3mIFfM/Cvatj3mIFfM?cls=10) in the provider, and users would need to receive API errors for feedback.

I think we cannot make a validator for scopes, but we can use CanonicalizeServiceScope in any ephemeral resources that handle scopes. Due to this, I think we should remove ServiceScopeValidator. What do you think?

Reply:
> From the discussion I linked it looks like [validating scopes isn't possible](https://chat.google.com/room/AAAA49vo02s/Cvatj3mIFfM/Cvatj3mIFfM?cls=10) in the provider, and users would need to receive API errors for feedback.
> 
> I think we cannot make a validator for scopes, but we can use CanonicalizeServiceScope in any ephemeral resources that handle scopes. Due to this, I think we should remove ServiceScopeValidator. What do you think?


Yeah it's unfortunate that some scopes don't expand to a uri. In this case we can just remove the ServiceScopeValidator to let the API handle it by returning an error